lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java

```
@@ -127,42 +129,58 @@ public FileConnector(Config config) throws ConnectorException {
+    for (String path : pathsToStorage) {
+      try {
+        URI newStorageURI = new URI(path);
+        storageURIs.add(newStorageURI);
+        log.debug("using path {} with scheme {}", path, newStorageURI.getScheme());
+      } catch (URISyntaxException e) {
+        throw new ConnectorException("Invalid path to storage: " + path, e);
+      }
     }
   }
 
   @Override
   public void execute(Publisher publisher) throws ConnectorException {
     try {
-      storageClient = StorageClient.create(storageURI, config);
-    } catch (Exception e) {
-      throw new ConnectorException("Error occurred while creating storage client.", e);
+      for (StorageClient client : storageClientMap.values()) {
+        client.init();
+      }
+    } catch (IOException e) {
+      throw new ConnectorException("Error initializing StorageClient.", e);
     }
 
     try {
-      storageClient.init();
-      TraversalParams params = new TraversalParams(storageURI, getDocIdPrefix(), fileOptions, filterOptions);
-      storageClient.traverse(publisher, params);
-    } catch (Exception e) {
-      throw new ConnectorException("Error occurred while initializing client or publishing files.", e);
+      for (URI pathToTraverse : storageURIs) {
+        try {
+          String clientKey = pathToTraverse.getScheme() != null ? pathToTraverse.getScheme() : "file";
+          StorageClient storageClient = storageClientMap.get(clientKey);
+
+          if (storageClient == null) {
+            log.warn("No StorageClient was available for {}, the path will be skipped. Did you include the necessary configuration?", pathToTraverse);
+            continue;
+          }
+
+          TraversalParams params = new TraversalParams(pathToTraverse, getDocIdPrefix(), fileOptions, filterOptions);
+          storageClient.traverse(publisher, params);
+        } catch (Exception e) {
+          throw new ConnectorException("Error occurred while initializing client or publishing files.", e);
+        }
```

It's a small detail, but wondering if we want to throw an exception if one path fails. I think the reason I'm leaning towards this is that, if an exception is thrown, the `StorageClient` itself could be in an unusable state, so if we are working with all Google Cloud paths, or something, this might not be very helpful.

---

I think it's ok to throw this exception
each traversal should be optimistic
so, if any traversal does throw an exception, that indicates a "significant" problem and we should bubble it up